### PR TITLE
refactor: preallocate package diff maps

### PR DESCRIPTION
**Pull request by lzap**
This is a wrap-up of my memory optimalisation research. I was mainly focusing on image building, this was not complete investigation of all parts of the application. Only what is on fire right now - image building.

In the first patch, I comment out pprof profiler, I am done with the process. Keeping the three lines in the codebase in case we need this again.

Second, it looks like the biggest offender during my (quite narrow) testing were package diff functions, while it could be optimized further, I am just preallocating maps to the specific size - this is always a good practice and this will save some copying.

The biggest issue is that kubernetes count cached pages into memory limit of a pod and while the application takes around 200-300 MB of memory when doing an image build, Linux filesystem cache goes up significantly (2-6 GB) which can ultimately lead to OOM kill of a pod. The third patch is an attempt to prevent that by using direct linux I/O kernel flag `O_DIRECT` which causes the read and write operations to bypass Linux kernel cache. I searched for only the relevant `.Open` and `os.Create` statements and added the flag there by changing this to the underlaying `os.FileOpen` call. This affects downloading and uploading of ISO files, tar extracting and S3 uploading, I did not do the change for irrelevant file operations (e.g. kickstart, tests, temporary files).

Most of this code will be removed once we implement the planned Pulp integration.

I noticed one file handler leak, fixed that as well.

### CODE REVIEW

File: pkg/services/devices.go [hunk 1]
```diff
@@ -378,7 +378,7 @@ func getPackageDiff(a, b []models.InstalledPackage) []models.InstalledPackage {
 
 func getVersionDiff(new, old []models.InstalledPackage) []models.InstalledPackage {
 	var diff []models.InstalledPackage
-	oldPkgs := make(map[string]models.InstalledPackage)
+	oldPkgs := make(map[string]models.InstalledPackage, len(old))
 	for _, pkg := range old {
 		oldPkgs[pkg.Name] = pkg
 	}
```
Comment (lzap) [inline pkg/services/devices.go:384]: This simply pre-allocates map so no copying is necessary - we know the length of the input slices in advance.

Turns out these functions are on the top of memory allocations, they do about 30% of all allocations of the application during my testing.

File: pkg/services/files.go [hunk 2]
```diff
@@ -79,7 +80,7 @@ func NewFilesService(log log.FieldLogger) FilesService {
 // GetFile returns the file given a path
 func (s *LocalFilesService) GetFile(path string) (io.ReadCloser, error) {
 	path = "/tmp/" + path
-	f, err := os.Open(filepath.Clean(path))
+	f, err := os.OpenFile(filepath.Clean(path), os.O_RDONLY|syscall.O_DIRECT, 0)
```
Comment (lzap) [inline pkg/services/files.go:83]: The rest is the same - changing `Open` or `Create` calls to `OpenFile` which allows passing flags. Double check if I did not make a mistake:

* `Open` should be `OpenFile(filename, os.O_RDONLY, 0)`
* `Create` should be `OpenFile(filename, O_RDWR|O_CREATE|O_TRUNC, 0666)` 

Taken directly from the Go codebase.

File: pkg/services/files/uploader.go [hunk 3]
```diff
@@ -201,10 +204,7 @@ func (u *S3Uploader) UploadFileWithACL(fname string, uploadPath string, acl stri
 		u.log.WithField("error", err.Error()).Error("Error uploading to AWS S3")
 		return "", err
 	}
-	if err := f.Close(); err != nil {
-		u.log.WithField("error", err.Error()).Error("Error closing file")
-		return "", err
-	}
+
```
Comment (lzap) [inline pkg/services/files/uploader.go:207]: This was a file handler leak in case the S3 PUT operation encountered an error.

**Comment by lzap** [comment]
Ah tests are failing with `invalid argument` since Linux direct I/O needs to be 512B aligned, okay. There are Go libraries for that but I was hoping only open flags would do the trick. Since most of the code is going away, I am dropping this commit from the PR. So it is only the two problems fixed now.

**Review by ezr-ondrej — DISMISSED** [review]
Nice finds :+1:

File: main.go [hunk 4]
```diff
@@ -126,10 +125,10 @@ func webRoutes(cfg *config.EdgeConfig) *chi.Mux {
 		http.ServeFile(w, r, cfg.OpenAPIFilePath)
 	})
 
-	// Non-production routes
-	if strings.Contains(config.Get().EdgeAPIBaseURL, "stage") {
-		route.Mount("/api/edge/debug", middleware.Profiler())
-	}
+	// Uncomment to enable profiling
+	//if strings.Contains(config.Get().EdgeAPIBaseURL, "stage") {
```
Comment (ezr-ondrej) [inline main.go:129]: linter does not like this. it's a two liner, I'd just delete it, maybe add it in `/docs/DEBBUGGING.md` as a tip instead if you want to go fancy.

Comment (ezr-ondrej) [reply main.go:129]: This needs to be fixed first :)

**Pull request merged**